src/java/org/apache/cassandra/service/accord/AccordJournal.java

```diff
@@ -492,54 +504,202 @@ public void forEach(Consumer<JournalKey> consumer)
+                    }
+
                     Journal.KeyRefs<JournalKey> ref = iter.next();
-                    key = ref.key();
-                    if (key.type != JournalKey.Type.COMMAND_DIFF)
+                    if (ref.key().type != COMMAND_DIFF)
                         continue;
 
+                    key = ref.key();
                     segments = journalTable.shouldIndex(key) ? ref.copyOfSegments() : null;
+                    break;
                 }
 
-                CommandStore commandStore = commandStores.forId(key.commandStoreId);
-                AccordCommandStoreLoader loader = (AccordCommandStoreLoader) commandStore.loader();
 
                 TxnId txnId = key.id;
                 Invariants.require(prev == null ||
                                    key.commandStoreId != prev.commandStoreId ||
                                    key.id.compareTo(prev.id) != 0,
                                    "duplicate key detected %s == %s", key, prev);
                 prev = key;
-
-                concurrency.acquireThrowUncheckedOnInterrupt(1);
+                commandParallelism.acquireThrowUncheckedOnInterrupt(1);
                 loader.load(txnId)
                       .map(route -> {
-                          if (segments != null)
+                          if (segments != null && route != null)
                           {
                               for (long segment : segments)
-                                  journalTable.safeNotify(index -> index.update(segment, key.commandStoreId, txnId, route));
+                                  journalTable.safeNotify(index -> index.update(segment, key.commandStoreId, txnId, (Route)route));
                           }
                           return null;
                       }).begin((success, fail) -> {
-                          concurrency.release(1);
+                          commandParallelism.release(1);
                           if (fail != null && !journal.handleError("Could not replay command " + txnId, fail))
                               abort.set(true);
                       });
+
+                return true;
+            }
+
+            @Override
+            public void close()
+            {
+                iter.close();
+            }
+        }
+
+        final Int2ObjectHashMap<ReplayStream> replayStreams = new Int2ObjectHashMap<>();
+        try
+        {
+            int cur = 0;
+            while (true)
+            {
+                if (abort.get())
+                    break;
+
+                if (cur == activeCommandStoreIds.size())
+                {
+                    if (activeCommandStoreIds.size() < commandStoreParallelism && !pendingCommandStores.isEmpty())
+                    {
+                        CommandStore next = pendingCommandStores.next();
+                        int id = next.id();
+                        activeCommandStoreIds.add(id);
+                        replayStreams.put(id, new ReplayStream(next));
```

Comment:
I think we need to pre-allocate all streams head of time to grab hold of static segments. This probably does not matter in test, but might have some effect in real word scenarios.

Reply:
Are you concerned they might get compacted? Or that they’ll read something they shouldn’t…?  I can reference the segments but wouldn’t want to open the iterators to avoid allocating all of the memory

Reply:
I actually don't think there's a need for the complexity here. I don't think we can write new data for command stores we haven't started replaying yet, and if the existing ones get compacted that also seems fine so long as we witness the newer ones....?

Reply:
Talked on slack: let's leave it as-is, since we can't pick up segments for _already active_ journals anyways, and other ones won't generate new work.